upstairs/src/impacted_blocks.rs

```diff
@@ -0,0 +1,275 @@
+        let bv = self.blocks.entry(extent_id).or_insert_with(|| {
+            let mut bv = BitVec::with_capacity(blocks_in_extent);
+            bv.resize(blocks_in_extent, false);
+            bv
+        });
+
+        bv.set(block.value as usize, true);
+    }
+
+    /// Return true if this list of impacted blocks overlaps with another.
+    pub fn conflicts(&self, other: &ImpactedBlocks) -> bool {
+        for shared_key in self
+            .blocks
+            .keys()
+            .filter(|key| other.blocks.contains_key(key))
+        {
+            // TODO any way to avoid this clone?
+            let bv = self.blocks[shared_key].clone();
+            if bv.bitand(&other.blocks[shared_key]).any() {
+                return true;
+            }
+        }
+
+        false
+    }
+
+    /// Returns (extent id, block) tuples
+    pub fn tuples(&self) -> Vec<(u64, Block)> {
+        let mut result = Vec::with_capacity(self.len());
+
+        let sorted_keys = self.blocks.keys().sorted();
+        for eid in sorted_keys {
+            let block_offsets =
+                self.blocks[eid].iter_ones().collect::<Vec<usize>>();
+            for i in block_offsets.iter().sorted() {
+                result
+                    .push((*eid, Block::new_with_ddef(*i as u64, &self.ddef)));
+            }
+        }
+
+        result
+    }
+
+    /// Returns the number of impacted blocks
+    pub fn len(&self) -> usize {
+        let mut len = 0;
+
+        for bv in self.blocks.values() {
+            len += bv.iter_ones().count();
+        }
+
+        len
+    }
+
+    #[allow(dead_code)]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+}
+
+/// Given an offset and number of blocks, return a list of individually impacted
+/// blocks:
+///
+///  |eid0                   |eid1
+///  |───────────────────────────────────────────────│
+///  ┌───────────────────────|───────────────────────┐
+///  │   |   |xxx|xxx|xxx|xxx|xxx|xxx|xxx|   |   |   │
+///  └───────────────────────|───────────────────────┘
+///          |--------------------------|
+///          offset                     offset + len
+///
+/// The example above would return a list of 7 impacted blocks.
+pub fn extent_from_offset(
+    ddef: RegionDefinition,
+    offset: Block,
+    num_blocks: Block,
+) -> ImpactedBlocks {
+    assert!(num_blocks.value > 0);
+    assert!(
+        (offset.value + num_blocks.value)
+            <= (ddef.extent_size().value * ddef.extent_count() as u64)
+    );
+    assert_eq!(offset.block_size_in_bytes() as u64, ddef.block_size());
+
+    let mut result = ImpactedBlocks::new(ddef);
+    let mut o: u64 = offset.value;
+    let mut blocks_left: u64 = num_blocks.value;
+
+    while blocks_left > 0 {
+        /*
+         * XXX We only support a single region (downstairs). When we grow to
+         * support a LBA size that is larger than a single region, then we
+         * will need to write more code. But - that code may live
```

Comment:
Since you are here already, just take out the "But - that code may ..." sentence as it makes no sense.

Reply:
addressed in faf6277